sql: lookup for ComparisonExpr ((column1)[void] IS DISTINCT FROM (NULL)[unknown])[bool]'s CmpOp failed

roachtest.sqlsmith/setup=empty/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5633774?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5633774?buildTab=artifacts#/sqlsmith/setup=empty/setting=no-mutations) on master @ [3c9b17113488d2ee6929936aa6ec48396f3ed71c](https://github.com/cockroachdb/cockroach/commits/3c9b17113488d2ee6929936aa6ec48396f3ed71c):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:265,sqlsmith.go:325,test_runner.go:896: error: pq: internal error: lookup for ComparisonExpr ((column1)[void] IS DISTINCT FROM (NULL)[unknown])[bool]'s CmpOp failed
		stmt:
		SELECT
			COALESCE(tab_115318.col_199168, NULL) AS col_199169
		FROM
			(VALUES ('':::VOID), (NULL), (NULL), ('':::VOID)) AS tab_115318 (col_199168)
		WHERE
			(NULL OR true)
		ORDER BY
			tab_115318.col_199168 ASC,
			tab_115318.col_199168,
			tab_115318.col_199168 DESC,
			tab_115318.col_199168 ASC,
			tab_115318.col_199168 DESC,
			tab_115318.col_199168;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Epic: CRDB-14914

Jira issue: CRDB-17240

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -2115,22 +2115,21 @@ func planProjectionOperators(
 		op, resultIdx, typs, err = planCastOperator(ctx, acc, typs, op, resultIdx, expr.ResolvedType(), t.ResolvedType(), factory, evalCtx)
 		return op, resultIdx, typs, err
 	case *tree.CoalesceExpr:
-		// We handle CoalesceExpr by planning the equivalent CASE expression,
-		// namely
+		// We handle CoalesceExpr by planning the equivalent CASE expression.
+		// Each WHEN condition is `IS DISTINCT FROM NULL` if the expression in the
+		// Coalesce allows that operation, otherwise, IS NOT NULL is used.
+		// For example,
+		//
 		//   CASE
 		//     WHEN CoalesceExpr.Exprs[0] IS DISTINCT FROM NULL THEN CoalesceExpr.Exprs[0]
-		//     WHEN CoalesceExpr.Exprs[1] IS DISTINCT FROM NULL THEN CoalesceExpr.Exprs[1]
+		//     WHEN CoalesceExpr.Exprs[1] IS NOT NULL THEN CoalesceExpr.Exprs[1]
 		//     ...
 		//   END
 		whens := make([]*tree.When, len(t.Exprs))
 		for i := range whens {
 			whens[i] = &tree.When{
-				Cond: tree.NewTypedComparisonExpr(
-					treecmp.MakeComparisonOperator(treecmp.IsDistinctFrom),
-					t.Exprs[i].(tree.TypedExpr),
-					tree.DNull,
-				),
-				Val: t.Exprs[i],
+				Cond: t.GetWhenCondition(i),
+				Val:  t.Exprs[i],
 			}
 		}
 		caseExpr, err := tree.NewTypedCaseExpr(

--- a/pkg/sql/logictest/testdata/logic_test/void
+++ b/pkg/sql/logictest/testdata/logic_test/void
@@ -40,3 +40,110 @@ query T
 SELECT crdb_internal.void_func()
 ----
 ·
+
+# Regression test for #83754. Postgres does not error out, but is inconsistent
+# in evaluation. For example, `SELECT ''::VOID IS DISTINCT FROM NULL::UNKNOWN;`
+# errors out, but `SELECT ''::VOID IS DISTINCT FROM NULL;` does not.
+# This is due to normalization into an IS NOT NULL op when one operand is NULL.
+# The NULL with type cast is not recognized as NULL.
+# We will stay consistent and have both cases error out to avoid different
+# behavior between normalized and unnormalized expressions.
+statement error pq: unsupported comparison operator: <void> IS DISTINCT FROM <unknown>
+SELECT ''::VOID IS DISTINCT FROM NULL
+
+statement ok
+SET vectorize=on
+
+# Regression test for #83754. Vectorized execution should not error out.
+query T
+SELECT
+  COALESCE(tab_115318.col_199168, NULL) AS col_199169
+FROM
+  (VALUES (''::VOID), (NULL), (NULL), (''::VOID)) AS tab_115318 (col_199168)
+ORDER BY
+  tab_115318.col_199168
+----
+NULL
+NULL
+·
+·
+
+# Regression test for #83754. Illegal type comparison should be caught in the
+# parser.
+statement error pq: incompatible NULLIF expressions: unsupported comparison operator: <void> = <unknown>
+SELECT NULLIF(tab_115318.a, tab_115318.b) AS col_199169
+FROM
+  (VALUES (''::VOID, NULL), (NULL, NULL)) AS tab_115318 (a, b)
+ORDER BY
+  tab_115318.a;
+
+statement error pq: incompatible NULLIF expressions: unsupported comparison operator: <void> = <void>
+SELECT NULLIF(tab_115318.a, tab_115318.b) AS col_199169
+FROM
+  (VALUES (''::VOID, ''::VOID), (NULL, NULL)) AS tab_115318 (a, b)
+ORDER BY
+  tab_115318.col_199168;
+
+# Regression test for #83754. Tuple type should be handled properly.
+query T
+SELECT
+  COALESCE(tab_115318.col_199168, NULL) AS col_199169
+FROM
+  (VALUES ((NULL, 1)), (NULL)) AS tab_115318 (col_199168)
+ORDER BY
+  tab_115318.col_199168
+----
+NULL
+(,1)
+
+statement ok
+SET vectorize=off
+
+# Regression test for #83754.
+# Non-vectorized and vectorized results should match.
+query T
+SELECT
+  COALESCE(tab_115318.col_199168, NULL) AS col_199169
+FROM
+  (VALUES (''::VOID), (NULL), (NULL), (''::VOID)) AS tab_115318 (col_199168)
+ORDER BY
+  tab_115318.col_199168
+----
+NULL
+NULL
+·
+·
+
+# Regression test for #83754.
+# Non-vectorized and vectorized results should match.
+statement error pq: incompatible NULLIF expressions: unsupported comparison operator: <void> = <unknown>
+SELECT NULLIF(tab_115318.a, tab_115318.b) AS col_199169
+FROM
+  (VALUES (''::VOID, NULL), (NULL, NULL)) AS tab_115318 (a, b)
+ORDER BY
+  tab_115318.a;
+
+# Regression test for #83754.
+# Non-vectorized and vectorized results should match.
+statement error pq: incompatible NULLIF expressions: unsupported comparison operator: <void> = <void>
+SELECT NULLIF(tab_115318.a, tab_115318.b) AS col_199169
+FROM
+  (VALUES (''::VOID, ''::VOID), (NULL, NULL)) AS tab_115318 (a, b)
+ORDER BY
+  tab_115318.col_199168;
+
+# Regression test for #83754.
+# Non-vectorized and vectorized results should match.
+query T
+SELECT
+  COALESCE(tab_115318.col_199168, NULL) AS col_199169
+FROM
+  (VALUES ((NULL, 1)), (NULL)) AS tab_115318 (col_199168)
+ORDER BY
+  tab_115318.col_199168
+----
+NULL
+(,1)
+
+statement ok
+RESET vectorize

--- a/pkg/sql/sem/tree/expr.go
+++ b/pkg/sql/sem/tree/expr.go
@@ -722,6 +722,28 @@ func (node *CoalesceExpr) Format(ctx *FmtCtx) {
 	ctx.WriteByte(')')
 }
 
+// GetWhenCondition builds the WHEN condition to use for the ith expression
+// inside the Coalesce.
+func (node *CoalesceExpr) GetWhenCondition(i int) (whenCond Expr) {
+	leftExpr := node.Exprs[i].(TypedExpr)
+	rightExpr := DNull
+	// IsDistinctFrom is listed as IsNotDistinctFrom in CmpOps.
+	_, ok :=
+		CmpOps[treecmp.IsNotDistinctFrom].LookupImpl(leftExpr.ResolvedType(), rightExpr.ResolvedType())
+	// If the comparison is legal, use IS NOT DISTINCT FROM NULL.
+	// Otherwise, use IS NOT NULL.
+	if ok {
+		whenCond = NewTypedComparisonExpr(
+			treecmp.MakeComparisonOperator(treecmp.IsDistinctFrom),
+			leftExpr,
+			rightExpr,
+		)
+		return whenCond
+	}
+	whenCond = NewTypedIsNotNullExpr(leftExpr)
+	return whenCond
+}
+
 // DefaultVal represents the DEFAULT expression.
 type DefaultVal struct{}
 

--- a/pkg/sql/sem/tree/type_check.go
+++ b/pkg/sql/sem/tree/type_check.go
@@ -1346,6 +1346,16 @@ func (expr *NullIfExpr) TypeCheck(
 	if err != nil {
 		return nil, decorateTypeCheckError(err, "incompatible NULLIF expressions")
 	}
+	leftType := typedSubExprs[0].ResolvedType()
+	rightType := typedSubExprs[1].ResolvedType()
+	_, ok := CmpOps[treecmp.EQ].LookupImpl(leftType, rightType)
+	if !ok {
+		op := treecmp.ComparisonOperator{Symbol: treecmp.EQ}
+		err = pgerror.Newf(
+			pgcode.UndefinedFunction, "unsupported comparison operator: <%s> %s <%s>",
+			leftType, op, rightType)
+		return nil, decorateTypeCheckError(err, "incompatible NULLIF expressions")
+	}
 
 	expr.Expr1, expr.Expr2 = typedSubExprs[0], typedSubExprs[1]
 	expr.typ = retType

